src/engine/TPA.cc

```diff
@@ -982,16 +983,101 @@ bool TPASplit::verifyExactPower(unsigned short power) const {
     return res == s_False;
 }
 
+vec<PTRef> TPABase::houdiniCheck(PTRef invCandidates, PTRef transition, SafetyExplanation::FixedPointType allignment) {
+    SMTConfig config;
+    MainSolver solver(logic, config, "Fixed-point checker");
+    solver.push();
+    auto candidates = topLevelConjuncts(logic, invCandidates);
+    if (allignment == SafetyExplanation::FixedPointType::RIGHT) {
+        solver.insertFormula(getNextVersion(transition));
+    } else {
+        if (allignment == SafetyExplanation::FixedPointType::LEFT) { solver.insertFormula(transition); }
+    }
+    solver.push();
+
+    // RIGHT:
+    //   rightInvariants /\ currentLevelTransition /\ getNextVersion(transition) =>
+    //     shiftOnlyNextVars(currentLevelTransition);
+    // LEFT:
+    //   leftInvariants /\ transition /\ getNextVersion(currentLevelTransition) =>
+    //     shiftOnlyNextVars(currentLevelTransition);
+    while (candidates.size() > 128) {
+        for (int i = candidates.size() - 1; i >= 1; i -= 2) {
+            PTRef n_f = logic.mkAnd(candidates[i], candidates[i - 1]);
+            candidates.pop();
+            candidates.pop();
+            candidates.push(n_f);
+            if (candidates.size() <= 128) { break; }
+        }
+    }
+    //    invCandidates.append(conjuncts);
+    //    Atr(x, x') /\ tr(x', x'') => Atr(x, x'')
+    //    or
+    //    Atr(x', x'') /\  tr(x, x') => Atr(x, x'')
+    //    Push transition once and for all
+    //    While loop externally, because we may drop smth important
+    PTRef goal = shiftOnlyNextVars(logic.mkAnd(candidates));
+
+    if (allignment == SafetyExplanation::FixedPointType::RIGHT) {
+        solver.insertFormula(logic.mkAnd({logic.mkAnd(candidates), logic.mkNot(goal)}));
+    } else {
+        if (allignment == SafetyExplanation::FixedPointType::LEFT) {
+            solver.insertFormula(logic.mkAnd({getNextVersion(logic.mkAnd(candidates)), logic.mkNot(goal)}));
+        }
+    }
+    while (solver.check() == s_True) {
+        for (int i = candidates.size() - 1; i >= 0; i--) {
+            PTRef cand = candidates[i];
+            solver.pop();
+            solver.push();
+            if (allignment == SafetyExplanation::FixedPointType::RIGHT) {
+                solver.insertFormula(logic.mkAnd({logic.mkAnd(candidates), logic.mkNot(shiftOnlyNextVars(cand))}));
+            } else {
+                if (allignment == SafetyExplanation::FixedPointType::LEFT) {
+                    solver.insertFormula(
+                        logic.mkAnd({getNextVersion(logic.mkAnd(candidates)), logic.mkNot(shiftOnlyNextVars(cand))}));
+                }
+            }
+            if (solver.check() == s_True) {
```

Comment:
I think you do not need to do this checks like this. I think it would be better to just get the model from the big query and remove all the candidates whose `shiftOnlyNextVars` version is falsified in this model.
Then repeat the big query with a smaller set of candidates.